pkg/snapshot/restorer/restorer.go

```diff
@@ -561,19 +619,31 @@ func applyEventsAndVerify(clientKV client.KVCloser, events []brtypes.Event, snap
 }
 
 // applyFirstDeltaSnapshot applies the events from first delta snapshot to etcd.
-func (r *Restorer) applyFirstDeltaSnapshot(clientKV client.KVCloser, snap brtypes.Snapshot) error {
+func (r *Restorer) applyFirstDeltaSnapshot(clientKV client.KVCloser, snap *brtypes.Snapshot) error {
 	r.logger.Infof("Applying first delta snapshot %s", path.Join(snap.SnapDir, snap.SnapName))
-	events, err := r.getEventsFromDeltaSnapshot(snap)
+
+	rc, err := r.store.Fetch(*snap)
 	if err != nil {
-		return fmt.Errorf("failed to read events from delta snapshot %s : %v", snap.SnapName, err)
+		return fmt.Errorf("failed to fetch delta snapshot %s from store : %v", snap.SnapName, err)
+	}
+
+	eventsData, err := r.readSnapshotContentsFromReadCloser(rc, snap)
+	if err != nil {
+		return fmt.Errorf("failed to read events data from delta snapshot %s : %v", snap.SnapName, err)
+	}
+
+	var events []brtypes.Event
+	if err = json.Unmarshal(eventsData, &events); err != nil {
+		return fmt.Errorf("failed to unmarshal events data from delta snapshot %s : %v", snap.SnapName, err)
 	}
 
 	// Note: Since revision in full snapshot file name might be lower than actual revision stored in snapshot.
-	// This is because of issue refereed below. So, as per workaround used in our logic of taking delta snapshot,
-	// latest revision from full snapshot may overlap with first few revision on first delta snapshot
+	// This is because of issue referred below. So, as per workaround used in our logic of taking delta snapshot,
+	// the latest revision from full snapshot may overlap with first few revision on first delta snapshot
 	// Hence, we have to additionally take care of that.
 	// Refer: https://github.com/coreos/etcd/issues/9037
-	ctx := context.TODO()
+	ctx, cancel := context.WithTimeout(context.TODO(), etcdDialTimeout)
```

Comment:
Using `context.TODO()` as the parent context for a timed context is fine. Because finally `ctx` will be a timed context.

Reply:
As per official documentation: https://pkg.go.dev/context#TODO
```
TODO returns a non-nil, empty Context. Code should use context.TODO when it's unclear which Context to use or it is not yet available (because the surrounding function has not yet been extended to accept a Context parameter).
```
when to use context.Background: https://pkg.go.dev/context#Background
I hope it should be clear now.